### PR TITLE
tetragon: Separate execve sensor into taill called programs

### PR DESCRIPTION
**Pull request by olsajiri**
The execve is getting more complex and on 4.19 it's reaching
the maximum program size.

Splitting execve sensor into 2 (so far) tail called programs and
adjusting TestLoad* tests.

Signed-off-by: Jiri Olsa <jolsa@kernel.org>

### CODE REVIEW

File: pkg/testutils/sensors/load.go [hunk 1]
```diff
@@ -55,6 +55,63 @@ func findProgram(cache []*prog, name string, typ ebpf.ProgramType, t *testing.T)
 	return nil
 }
 
+func mergeSensorMaps(maps1, maps2 []SensorMap, progs1, progs2 []SensorProg, t *testing.T) ([]SensorMap, []SensorProg) {
```
Comment (kkourt) [inline pkg/testutils/sensors/load.go:59]: I guess this function is only intended for  testing?

If so, let's `t *testing.T`  place as the first argument.
Otherwise, let's avoid `t.Fatalf()` and have it return an error.

File: pkg/testutils/sensors/load.go [hunk 2]
```diff
@@ -112,8 +112,33 @@ func mergeSensorMaps(maps1, maps2 []SensorMap, progs1, progs2 []SensorProg, t *t
 	return mapsReturn, progsReturn
 }
 
+func mergeInBaseSensorMaps(sensorMaps []SensorMap, sensorProgs []SensorProg, t *testing.T) ([]SensorMap, []SensorProg) {
```
Comment (kkourt) [inline pkg/testutils/sensors/load.go:116]: same with `mergeSensorMaps`: let's add the t *testing.T arg first.
Ideally, all test functions should be in _test.go files, but not sure if that's possible.

File: pkg/testutils/sensors/load.go [hunk 3]
```diff
@@ -132,6 +134,12 @@ func mergeInBaseSensorMaps(sensorMaps []SensorMap, sensorProgs []SensorProg, t *
 		SensorMap{Name: "execve_val", Progs: []uint{2}},
 	}
 
+	if kernels.EnableLargeProgs() {
+		baseMaps = append(baseMaps, SensorMap{Name: "tcpmon_map", Progs: []uint{0, 1, 2, 3}})
+	} else {
+		baseMaps = append(baseMaps, SensorMap{Name: "tcpmon_map", Progs: []uint{1, 2, 3}})
+	}
+
```
Comment (kkourt) [inline pkg/testutils/sensors/load.go:142]: Why is this needed?

My reading is:
Program 0 is event_execve, but I don't see how event_execve will ever need tcpmon_map since we seem to be alwats tail-calling into the program that sends the event (execve_send). Am I missing something?

File: pkg/testutils/sensors/load.go (hunk 1 above)
Comment (olsajiri) [reply pkg/testutils/sensors/load.go:59]: yes, it's just for testing to merge tests maps, I'll move the 't' arg to the first position

File: pkg/testutils/sensors/load.go (hunk 2 above)
Comment (olsajiri) [reply pkg/testutils/sensors/load.go:116]: yes, will move the arg

it's all under `pkg/testutils/sensors/load.go` so testing related

File: pkg/testutils/sensors/load.go (hunk 3 above)
Comment (olsajiri) [reply pkg/testutils/sensors/load.go:142]: there's the data interface that sends events and is part of event_execve,
but it's only compiled in for large programs.. it's still not there for 4.19

there's PR for that alreay, but I wanted to split the execve event first

**Review by kkourt — APPROVED** [review]

File: bpf/lib/process.h [hunk 4]
```diff
@@ -262,6 +262,10 @@ struct msg_execve_event {
 		struct msg_process process;
 		char buffer[PADDED_BUFFER];
 	};
+	/* below fields are not part of the event, serve just as
+	 * heap for execve programs
+	 */
+	__u32 binary;
```
Comment (tixxdz) [inline bpf/lib/process.h:268]: Wouldn't be possible to create a new struct keep it local to bpf_execve_event.c file that has both msg_execve_event and the binary embedded inside?

Just to keep msg_execve_event clean and for its purpose and maybe it reduce errors or confusions, yeh you commented it ;-) but we may also start doing same in future and it can get ugly !

**Review by tixxdz — APPROVED** [review]
Thank you @olsajiri this solves my problems too for 4.19 and cgroups ;-)

Approved with one comment, I think we shouldn't touch the struct but your call!

File: bpf/lib/process.h (hunk 4 above)
Comment (tixxdz) [reply bpf/lib/process.h:268]: I mark this resolved since we already do this for kprobes.

File: bpf/process/bpf_execve_event.c [hunk 5]
```diff
@@ -185,10 +188,37 @@ event_execve(struct sched_execve_args *ctx)
 	fileoff = ctx->filename & 0xFFFF;
 	binary = event_filename_builder(ctx, execve, pid, EVENT_EXECVE, binary,
 					(char *)ctx + fileoff);
+	event->binary = binary;
```
Comment (jrfastab) [inline bpf/process/bpf_execve_event.c:191]: this reads a bit strange I'm not sure why

event->binary = event_filename_builder(...)

is not used.

Comment (olsajiri) [reply bpf/process/bpf_execve_event.c:191]: ugh, will fix that in another PR, thanks
